matchpygen/example.cpp

```diff
@@ -0,0 +1,132 @@
+#include <iostream>
+#include <tuple>
+#include <map>
+#include <string>
+#include <deque>
+
+#include <symengine/basic.h>
+#include <symengine/pow.h>
+
+using namespace std;
+using namespace SymEngine;
+
+typedef map<string, RCP<const Basic>> Substitution2;
+
+int try_add_variable(Substitution2 &subst, string variable_name,
+                     RCP<const Basic> &replacement)
+{
+    if (subst.find(variable_name) == subst.end()) {
+        subst[variable_name] = replacement;
+        // this[variable_name] = replacement; //.copy() if
+        // isinstance(replacement, Multiset) else replacement
+    } else {
+        // Node &existing_value = *(this->find(variable_name));
+        /*
+        if (isinstance(existing_value, tuple)) {
+            if (isinstance(replacement, Multiset)) {
+                if (Multiset(existing_value) != replacement){
+                    return -1;
+                }
+            } else if (replacement != existing_value) {
+                return -1;
+            }
+            else if (isinstance(existing_value, Multiset)) {
+                if (!isinstance(replacement, (tuple, list, Multiset))) {
+                    return -1;
+                }
+                auto &compare_value = Multiset(replacement);
+                if (existing_value == compare_value){
+                    if (! isinstance(replacement, Multiset)) {
+                        this->insert(make_pair(variable_name, replacement));
+                    } else {
+                        return -1;
+                    }
+                } else if (replacement != existing_value){
+                    return -1;
+                }
+            }
+        }
+        */
+    }
+    return 0;
+}
+
+typedef deque<RCP<const Basic>> Deque;
+
+Deque get_deque(RCP<const Basic> expr)
+{
+    Deque d;
+    for (RCP<const Basic> i : expr->get_args()) {
+        d.push_back(i);
+    }
+    return d;
+}
+
+tuple<int, Substitution2> match_root(RCP<const Basic> subject)
```

Comment:
I think it's ok to use `std::variant` here, if that's what is intended.

For older compilers, we can provide a compatible implementation, say using Boost variant, i.e., by having `SymEngine::variant`, that will be equal to `std::variant` for compilers that support it, and `boost::variant` for older compilers. For now just use `std::variant`.